### PR TITLE
release: detect if we're running on core desktop

### PR DESCRIPTION
**Pull request by robert-ancell**

### CODE REVIEW

File: release/release.go
Comment (mvo5) [inline release/release.go:180]: I wonder if /etc/os-release might be a slightly better place? We could use `ID=ubuntu-core\nVARIANT_ID=ubuntu-core-desktop`  [1] or something similar?  

[1] https://www.freedesktop.org/software/systemd/man/os-release.html

Comment (robert-ancell) [reply release/release.go:180]: The `snap.yaml` check was recommended by @jhenstridge, at least for a first implementation. I think we ultimately need a specific flag, so any future core desktop can easily opt-in to this behaviour.

I think there's three IDs we need, as Ubuntu Core Desktop is:
- A specific OS
- A Core Desktop OS
- A version of Ubuntu Core

I think this would be represented as (might require snapd to be updated to check ID_LIKE):
- `ID=ubuntu-core-desktop\nID_LIKE=ubuntu-core`
- `ID=lubuntu-core-desktop\nID_LIKE=ubuntu-core ubuntu-core-desktop`

Comment (jhenstridge) [reply release/release.go:180]: My thoughts were that it was more important to get the `release.OnCoreDesktop` variable available to unblock other work, with the understanding that we could change how we determine the value of OnCoreDesktop is determined later.

With that said, the use of `VARIANT_ID` does sound like a reasonable fit for what we're trying to do. Looking at the spec, the variant ID is effectively private to the the distro ID, so we could just do `ID=ubuntu-core` and `VARIANT_ID=desktop`.

What we are building mostly behaves like regular Ubuntu Core, so I don't think we'd want to change the distro ID.

File: daemon/api_general.go
Comment (pedronis) [inline daemon/api_general.go:138]: it feels like ultimately this belong more under os-release than here, we should probably not expose it here until we are sure of the way forward

File: release/release.go
Comment (pedronis) [inline release/release.go:178]: this does not follow go conventions for docs, it should start with the name of the functions, given that is not exported I we are not super picky about it but then we tend to make our one-off comments all lowercase without a dot at the end.

File: release/release.go
Comment (pedronis) [inline release/release.go:174]: all the code using here needs some XXX saying that we hope to have a more stable implementation of this as checking for exactly core22-desktop is fragile going forward

File: release/release.go
Comment (pedronis) [reply release/release.go:180]: as I said eleswhere this means that ideally this should also get reflected into ReleaseInfo

File: release/release.go
Comment (pedronis) [inline release/release.go:218]: this as convenience for own code makes sense though.

**Review by pedronis — COMMENTED** [review]
some comments

File: release/release.go
Comment (robert-ancell) [reply release/release.go]: Changed to using VARIANT_ID

File: daemon/api_general.go
Comment (robert-ancell) [reply daemon/api_general.go:138]: Regardless of where the information comes from having it explicitly stated here is useful so other code e.g. snapd-desktop-integration can match the behaviour of snapd.

Comment (jhenstridge) [reply daemon/api_general.go:138]: At present snapd-desktop-integration has no access to the `/v2/system-info` API endpoint, so this can't help there. And if there was an area where we wanted snapd-desktop-integration to change its behaviour to match snapd, it might be better to have snapd express that policy directly.

It looks like the only place `on-classic` is used at the moment is in formatting the `snap version` output (deciding whether to show the distro ID and version ID or not). If we wanted to add more information to that command, the question is whether we'd want to show `on-core-desktop`, or just the release variant.

File: daemon/api_general.go
Comment (jhenstridge) [inline daemon/api_general.go:136]: As there is a new JSON exported field in this struct, it'd be good to update `OSRelease` in `client/client.go` to decode it.

If we keep `on-core-desktop` below, it should also be exposed in `SysInfo` from that file too.

**Review by jhenstridge — APPROVED** [review]
The release package changes look good to me.

I'm not so sure of the utility of adding `on-core-desktop` to the system-info API though: it doesn't hurt, but I'm also not sure whether we have a use case for it. I'd defer to @pedronis on that part.

File: daemon/api_general.go
Comment (pedronis) [reply daemon/api_general.go:138]: as I said my preference is to leave this out, at least until we understand more the behavior differences in snapd, but also as @jhenstridge mentioned this is not really available to snap-desktop-integration. It will be anyway reflected in the variant-id under os-release

File: release/release.go
Comment (pedronis) [inline release/release.go:222]: s/onClassic/onCoreDesktop/ ?

File: release/release.go
Comment (pedronis) [inline release/release.go:101]: this bit needs testing

**Review by pedronis — COMMENTED** [review]
changes to release.go look good, some comments

**Review by pedronis — APPROVED** [review]
thanks

**Review by mvo5 — APPROVED** [review]
Nice!

**Pull request merged**
